src/main/java/org/broadinstitute/hellbender/utils/locusiterator/ReadStateManager.java

```diff
@@ -0,0 +1,257 @@
+        if ( isEmpty() ) {
+            // there are no reads here, so our next state is the next read in the stream
+            firstContig = iterator.peek().getContig();
+            firstAlignmentStart = iterator.peek().getStart();
+        } else {
+            // there's a read in the system, so it's our targeted first read
+            final AlignmentStateMachine firstState = getFirst();
+            firstContig = firstState.getContig();
+            // note this isn't the alignment start of the read, but rather the alignment start position
+            firstAlignmentStart = firstState.getGenomePosition();
+        }
+
+        while ( iterator.hasNext() && readStartsAtCurrentPosition(iterator.peek(), firstContig, firstAlignmentStart) ) {
+            submitRead(iterator.next());
+        }
+
+        samplePartitioner.doneSubmittingReads();
+
+        for (final String sample : samples) {
+            final Collection<GATKRead> newReads = samplePartitioner.getReadsForSample(sample);
+
+            // if we're keeping reads, take the (potentially downsampled) list of new reads for this sample
+            // and add to the list of reads.  Note this may reorder the list of reads someone (it groups them
+            // by sample, but it cannot change their absolute position on the genome as they all must
+            // start at the current location
+            if ( keepSubmittedReads ) {
+                submittedReads.addAll(newReads);
+            }
+
+            final PerSampleReadStateManager statesBySample = readStatesBySample.get(sample);
+            addReadsToSample(statesBySample, newReads);
+        }
+
+        samplePartitioner.reset();
+    }
+
+    /**
+     * Add a read to the sample partitioner, potentially adding it to all submitted reads, if appropriate
+     * @param read a non-null read
+     */
+    void submitRead(final GATKRead read) {
+        samplePartitioner.submitRead(read);
+    }
+
+    /**
+     * Transfer current list of submitted reads, clearing old list
+     *
+     * Takes the maintained list of submitted reads, and transfers it to the caller of this
+     * function.  The old list of set to a new, cleanly allocated list so the caller officially
+     * owns the list returned by this call.  This is the only way to clear the tracking
+     * of submitted reads, if enabled.
+     *
+     * How to use this function:
+     *
+     * while ( doing some work unit, such as creating pileup at some locus ):
+     *   interact with ReadStateManager in some way to make work unit
+     *   readsUsedInPileup = transferSubmittedReads)
+     *
+     * @throws UnsupportedOperationException if called when keepSubmittedReads is false
+     *
+     * @return the current list of submitted reads
+     */
+    public List<GATKRead> transferSubmittedReads() {
+        if ( ! keepSubmittedReads ) throw new IllegalArgumentException("cannot transferSubmittedReads if you aren't keeping them");
```

Comment:
`IllegalArgumentException` makes no sense here given that this method takes no arguments :) Change to `IllegalStateException`